### PR TITLE
security(tools): restrict PromQL/LogQL queries to namespace-scoped patterns 🔐

### PR DESCRIPTION
**Pull request by Klazomenai**
## Summary

Restricts Lookout's `prometheus_query` and `loki_query` tools to namespace-scoped queries per bridge#80. Every vector selector (PromQL) and stream selector (LogQL) in every query MUST constrain the `namespace` label to an allowlisted value, or the query is rejected before the HTTP request is built.

Previously, these tools passed user-supplied queries straight through to the backend — an LLM (or prompt-injected LLM) could have issued `up` or `{}` and retrieved cluster-wide metrics/logs including series from namespaces containing secrets.

## Implementation

- **PromQL authorizer** (`internal/tools/lookout/authorize_promql.go`, 201 lines): parses the query with `github.com/prometheus/prometheus/promql/parser`, walks every `VectorSelector` in the AST, and checks its `namespace` label matcher.
- **LogQL authorizer** (`internal/tools/lookout/authorize_logql.go`, 133 lines): extracts balanced `{...}` stream selectors with a string-aware depth counter (to correctly handle regex values like `"foo{1,3}"` inside quoted matcher strings), then reuses the Prometheus label-matcher parser for each. This avoids importing the entire `grafana/loki` codebase for what reduces to a label-matcher syntax.
- **`NamespaceAllowlist`** (`internal/tools/lookout/allowlist.go`): small value type, parses the `LOOKOUT_NAMESPACE_ALLOWLIST` env var (comma-separated), stores values as a set, provides `Contains` + sorted `Names()` for logging.
- **`cmd/bridge/main.go`**: fails closed — when `LOOKOUT_NAMESPACE_ALLOWLIST` is empty, both tools register as stubs even if `PROMETHEUS_URL` / `LOKI_URL` are set. Matches the `CHIPS_REPO_ALLOWLIST` precedent.

## Rejection rules

The authorizer rejects queries that:

- Lack a `namespace` matcher on any selector (naked `up`, `{__name__="up"}`, `{app="bridge"}`)
- Use a negative matcher (`namespace!="x"` or `namespace!~"x.*"`) — these grant visibility to everything except the named value
- Use regex with wildcards, character classes, repetition, flags, or any non-literal construct (`.*`, `.+`, `ma.*`, `[a-z]+`, `matrix{1,10}`, `(?i)matrix`, `matrix|.*`)
- Use `MatchEqual` with a value not in the allowlist
- Use a regex literal alternation where ANY alternative is not allowlisted

The authorizer accepts:

- `namespace="matrix"` (simple equality, allowlisted)
- `namespace=~"matrix"` (regex literal)
- `namespace=~"matrix|argocd"` or `namespace=~"(matrix|argocd)"` (anchored alternation, all allowlisted)
- `sum(up{namespace="matrix"}) without (namespace)` — aggregations that drop the namespace from the RESULT are fine because the underlying vector selectors are still filtered

## Test coverage

- `authorize_promql_test.go`: 24 tests covering happy paths, naked queries, wildcard escapes, regex alternations, negative matchers, parse errors, fail-closed paths, subqueries, nested aggregations
- `authorize_logql_test.go`: 21 tests covering stream selector extraction (including brace-inside-quoted-string edge cases), multi-stream queries, line-filter-regex-with-braces, unterminated strings, fail-closed paths
- `allowlist_test.go`: 9 tests for the `NamespaceAllowlist` value type (whitespace trimming, nil-safe methods, empty-string defense)
- Existing `lookout_test.go` updated: every constructor call-site now passes a test allowlist; every query prefixed with `namespace="matrix"`

Package coverage: **86.4%**. Full suite passes with `-tags goolm -race`.

## Configuration

New env var: `LOOKOUT_NAMESPACE_ALLOWLIST` — comma-separated list of allowed namespace names. Whitespace around entries is trimmed; empty entries discarded. When unset or empty, the Lookout tools are stubbed regardless of `PROMETHEUS_URL` / `LOKI_URL`. Example for AKeyRA:

```yaml
env:
  - name: LOOKOUT_NAMESPACE_ALLOWLIST
    value: matrix,argocd,monitoring,vault,external-secrets
```

## Architectural notes

Per the tooling synthesis that informed this design, the allowlist is a struct field on each tool (not a package global), so when M4 per-crew pods land (#42, #50, #52), each crew pod can be started with a different `LOOKOUT_NAMESPACE_ALLOWLIST` without code changes.

The authorizer runs **before** the HTTP request is built — output truncation in the sandbox cannot undo data exfiltration, so the check must be pre-request.

## Dependencies added

- `github.com/prometheus/prometheus v0.311.2` (for `promql/parser`)
- 10 indirect deps (`prometheus/common`, `prometheus/client_golang`, `grafana/regexp`, etc.)

Binary size impact to be measured on the next release build.

Refs #80

## Test plan

- [x] `go test -tags goolm ./...` — all packages pass
- [x] `go test -tags goolm -race ./...` — no data races
- [x] `go test -tags goolm -cover ./internal/tools/lookout/...` — 86.4%
- [x] `go vet -tags goolm ./...` — clean
- [x] `gofmt -l` — clean
- [x] Review
- [ ] Deploy to AKeyRA with `LOOKOUT_NAMESPACE_ALLOWLIST=matrix,argocd,monitoring,vault,external-secrets` and `PROMETHEUS_URL=...` to exercise live; confirm `lookout: prometheus_query registered` in startup logs instead of stub

### CODE REVIEW

File: internal/tools/lookout/authorize_promql.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+// contains wildcards, character classes, repetition).
+func extractLiterals(re *syntax.Regexp) ([]string, error) {
+	// Prometheus anchors regex matchers at both ends. The parser returns a
+	// tree that may be wrapped in OpCapture; unwrap it.
+	re = unwrap(re)
+
+	switch re.Op {
+	case syntax.OpLiteral:
+		return []string{string(re.Rune)}, nil
+
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply internal/tools/lookout/authorize_promql.go]: Good catch — verified empirically (see commit 8b37c0e):

```
pattern="(?i)matrix"  Op=Literal  Rune="MATRIX"  Flags=0xd5  FoldCase=true
```

The previous test `TestAuthorizePromQL_RegexWithFlags` passed only by accident — the normalised "MATRIX" rune happened not to literal-match the lowercase allowlist entry. If anyone ever allowlisted an uppercase value, the case-fold would silently widen the backend match.

Fix adds `checkNoUnsafeFlags` in authorize_promql.go that walks the entire regex tree and rejects any node with `FoldCase | DotNL | NonGreedy | Literal`. Recursion is mandatory because the parser propagates the flag onto leaf literals while the parent node may look clean — confirmed by probing `matrix(?i)argocd`, which yields:

```
Op=Concat         Flags=0x0     (clean!)
  Op=Literal "matrix"  Flags=0xd4  (clean)
  Op=Literal "ARGOCD"  Flags=0xd5  (FoldCase)
```

New tests cover `(?i)` against uppercase allowlist, `(?i:sub)rest`, `matrix(?i)argocd`, and `(?s)` DotNL.

File: go.mod [hunk 2]
```diff
@@ -13,27 +13,40 @@ require (
...
 	github.com/mattn/go-sqlite3 v1.14.34 // indirect
+	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
 	github.com/ncruces/go-strftime v1.0.0 // indirect
 	github.com/petermattis/goid v0.0.0-20260226131333-17d1149c6ac6 // indirect
+	github.com/prometheus/client_golang v1.23.2 // indirect
+	github.com/prometheus/client_model v0.6.2 // indirect
+	github.com/prometheus/common v0.67.5 // indirect
+	github.com/prometheus/procfs v0.16.1 // indirect
+	github.com/prometheus/prometheus v0.311.2 // indirect
 	github.com/remyoudompheng/bigfft v0.0.0-20230129092748-24d4a6f8daec // indirect
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply go.mod:36]: Fixed in commit 8b37c0e — `go mod tidy` moved `github.com/prometheus/prometheus` from the indirect block into the direct `require` block alongside the existing direct deps, plus a few other cleanups (`github.com/kr/text` added as test indirect from the newer dep closure).

File: internal/tools/lookout/authorize_promql.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+// unsafeRegexFlags are the inline flag directives a caller must NOT be able to
+// introduce via an in-regex `(?flags)` construct. Case-folding lets "(?i)matrix"
+// widen the match beyond "matrix" at the backend even though our literal
+// extraction collapses it to a single rune-string; DotNL and NonGreedy are
+// meaningless for literals but rejecting them pre-empts accidental
+// interactions with future regex-construct support.
+const unsafeRegexFlags = syntax.FoldCase | syntax.DotNL | syntax.NonGreedy | syntax.Literal
+
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply internal/tools/lookout/authorize_promql.go]: Good catch — verified empirically:

```
pattern="matrix"     → Flags=0xD4  (Perl default)
pattern="(?m)matrix" → Flags=0xC4  (Perl minus OneLine)
pattern="(?U)matrix" → Flags=0xF4  (Perl plus NonGreedy)
```

`(?m)` CLEARS OneLine rather than setting a new bit — a flag bitmask check like `unsafeRegexFlags` simply cannot detect it (absence ≠ presence).

Fixed in commit 395a858 by switching to a pre-parse string-level scanner (`flagDirectiveRe = \(\?[a-zA-Z\-]+[:)]`) that rejects any Perl-style inline flag directive before the regex reaches `syntax.Parse`. The bitmask walker is kept as defence in depth. The scanner deliberately allows `(?:...)` (non-capturing group, no flag letters) and `(?P<name>...)` (Go named-capture, `P` followed by `<`) — both have explicit allow tests to pin the false-positive guard.

New rejection tests: `(?m)`, `(?U)`, `(?-i)`.

File: cmd/bridge/main.go [hunk 4]
```diff
@@ -108,19 +108,34 @@ func main() {
...
+	promURL := os.Getenv("PROMETHEUS_URL")
+	lokiURL := os.Getenv("LOKI_URL")
+	switch {
+	case promURL == "":
 		toolReg.Register(tools.NewStubTool("prometheus_query", "Query Prometheus metrics (PROMETHEUS_URL not set)"))
 		slog.Info("lookout: prometheus_query registered as stub (PROMETHEUS_URL not set)")
+	case lookoutAllowlist.Len() == 0:
+		toolReg.Register(tools.NewStubTool("prometheus_query", "Query Prometheus metrics (LOOKOUT_NAMESPACE_ALLOWLIST not set)"))
+		slog.Warn("lookout: prometheus_query registered as stub (LOOKOUT_NAMESPACE_ALLOWLIST not set)")
+	default:
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply cmd/bridge/main.go:125]: Fixed in commit 395a858. The stub description and the `slog.Warn` message now both say "LOOKOUT_NAMESPACE_ALLOWLIST not set or empty", which correctly covers `ParseNamespaceAllowlist` dropping whitespace-only and empty entries.

File: cmd/bridge/main.go [hunk 5]
```diff
@@ -108,19 +108,34 @@ func main() {
...
-		toolReg.Register(lookouttools.NewLokiQueryTool(lokiURL, lookouttools.DefaultHTTPClient()))
-		slog.Info("lookout: loki_query registered", "url", lokiURL)
-	} else {
+	switch {
+	case lokiURL == "":
 		toolReg.Register(tools.NewStubTool("loki_query", "Query Loki logs (LOKI_URL not set)"))
 		slog.Info("lookout: loki_query registered as stub (LOKI_URL not set)")
+	case lookoutAllowlist.Len() == 0:
+		toolReg.Register(tools.NewStubTool("loki_query", "Query Loki logs (LOOKOUT_NAMESPACE_ALLOWLIST not set)"))
+		slog.Warn("lookout: loki_query registered as stub (LOOKOUT_NAMESPACE_ALLOWLIST not set)")
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply cmd/bridge/main.go]: Same fix as the Prometheus branch in commit 395a858 — stub message now reads "LOOKOUT_NAMESPACE_ALLOWLIST not set or empty".
